mitxgraders/helpers/specify_domain.py

```diff
@@ -0,0 +1,241 @@
+    '21th'
+    """
+    first_three = {1: '1st', 2: '2nd', 3: '3rd'}
+    return first_three.get(n, '{0}th'.format(n))
+
+def get_description(obj):
+    """
+    Gets a student-facing description of obj.
+
+    Numbers return scalar:
+    >>> get_description(5)
+    'scalar'
+
+    MathArrays return their own description:
+    >>> get_description(MathArray([1, 2, 3]))
+    'vector of length 3'
+
+    Other objects return their class name:
+    >>> get_description("puppy")
+    'str'
+    """
+    if isinstance(obj, Number):
+        return 'scalar'
+    elif isinstance(obj, MathArray):
+        return obj.description
+    else:
+        return obj.__class__.__name__
+
+class DomainError(StudentFacingError):
+    """
+    Raised when a function has domain error.
+    """
+
+def number_validator(obj):
+    """
+    Voluptuous validator to test that obj is number.
+
+    >>> number_validator(5) == 5
+    True
+    >>> number_validator(MathArray(5)) == 5
+    True
+
+    Provides a useful error message:
+    >>> number_validator(MathArray([1, 2, 3]))
+    Traceback (most recent call last):
+    Invalid: received a vector of length 3, expected a scalar
+    >>> number_validator([1, 2, 3])
+    Traceback (most recent call last):
+    Invalid: received a list, expected a scalar
+
+    """
+    if isinstance(obj, Number):
+        return obj
+    elif is_scalar_matharray(obj):
+        return obj.item()
+
+    raise Invalid("received a {0}, expected a scalar".format(get_description(obj)))
+
+
+def make_shape_validator(shape):
+    """
+    Returns a voluptuous validator that tests whether argument is a MathArray
+    with specified shape.
+
+    Arguments:
+    =========
+        shape (tuple): A numpy shape tuple
+
+    Usage:
+    ======
+    >>> validate_vec4 = make_shape_validator((4,))
+    >>> vec4 = MathArray([1, 2, 3, 4])
+    >>> validate_vec4(vec4)
+    MathArray([1, 2, 3, 4])
+
+    Provides useful error messages if obj is a number or MathArray:
+    >>> validate_vec4(MathArray([[1, 2, 3], [4, 5, 6]]))
+    Traceback (most recent call last):
+    Invalid: received a matrix of shape (rows: 2, cols: 3), expected a vector of length 4
+    >>> validate_vec4('cat')
+    Traceback (most recent call last):
+    Invalid: received a str, expected a vector of length 4
+    """
+    def shape_validator(obj):
+        if isinstance(obj, MathArray) and obj.shape == shape:
+            return obj
+        raise Invalid("received a {received}, expected a {expected}"
+            .format(received=get_description(obj), expected=MathArray.get_description(shape)))
+
+    return shape_validator
+
+def has_shape(shape):
+    if shape == (1,):
+        return number_validator
+    return make_shape_validator(shape)
+
+# how authors specify domain shape
+SCHEMA_SHAPE = Schema(Any(
+    All(Positive(int), lambda x: tuple([x])),
+    All([Positive(int)], Coerce(tuple)),
+    msg='Expected shape specification to be a positive integer or list of positive integers'
```

Comment:
Can authors supply a numpy shape tuple?

Reply:
Not currently, and I think we should not allow it, though I am somewhat torn.

- numpy uses tuple for storing shape (I suspect because they don't want you mutating the shape property of an array). So it would be logical to use tuples for shape.
- On the other hand, in our configuration schemas, we uses lists for *almost* everything that requires enumeration. The only exception I can think of is using a tuple for enumerating multiple answers.

  In particular, suppose we use tuple for shape. We should be consistent, so then matrix sampling sets become:
  ```python
  RealMatrices(shape=(3, 2), norm=[1, 5])
  ```
  What the heck? Why is one a tuple and one is a list? That seems confusingly arbitrary!
- We could allow specifying shape as a list *or* tuple. But that's a pain from voluptuous's perspective, and makes documentation (a little bit) awkward.